src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,908 @@
+                            float rank(0.0), coverage(0.0);
+                            if (m3hash != 0)
+                            {
+                                rank = hash_to_float(m3hash);
+                                coverage = ranked_vector[i].first / total_weight;
+                            }
+                            pixel_values.push_back(rank);
+                            pixel_values.push_back(coverage);
+                        }
+
+                        const size_t num_channels = (m_num_layers * 2) + 3;
+                        for (size_t i = ranked_vector.size(); i < num_channels; ++i)
+                            pixel_values.push_back(0.0f);
+
+                        m_aov_image.set_pixel(rx, ry, pixel_values.data());
+                    }
+                }
+            }
+        }
+
+        void on_sample_begin(
+            const PixelContext&         pixel_context) override
+        {
+        }
+
+        void on_sample_end(
+            const PixelContext&         pixel_context) override
+        {
+        }
+
+        void on_pixel_begin(
+            const Vector2i&             pi) override
+        {
+        }
+
+        void on_pixel_end(
+            const Vector2i&             pi) override
+        {
+        }
+
+        void write(
+            const PixelContext&         pixel_context,
+            const ShadingPoint&         shading_point,
+            const ShadingComponents&    shading_components,
+            const AOVComponents&        aov_components,
+            ShadingResult&              shading_result) override
+        {
+            uint32 m3hash = 0;
+            string obj_name;
+            if (shading_point.hit_surface())
+            {
+                switch (m_layer_type)
+                {
+                  case CryptomatteAOV::CryptomatteType::ObjectNames:
+                    obj_name = shading_point.get_object().get_name();
+                    break;
+
+                  case CryptomatteAOV::CryptomatteType::MaterialNames:
+                    {
+                      const auto* obj_material = shading_point.get_material();
+                      if (obj_material != nullptr)
+                          obj_name = obj_material->get_name();
+                    }
+                    break;
+
+                  default:
+                    break;
+                }
+
+                MurmurHash3_x86_32(reinterpret_cast<const unsigned char*>(obj_name.c_str()), static_cast<int>(obj_name.size()), 0, &m3hash);
+            }
+
+            const Vector2i pixel_pos = pixel_context.get_pixel_coords();
+            const Vector2d sample_pos = pixel_context.get_sample_position();
+            int tile_pixel_x = pixel_pos.x - m_tile_origin_x;
+            int tile_pixel_y = pixel_pos.y - m_tile_origin_y;
+
+            m_tile_name_map[make_pair(m_tile_origin_x, m_tile_origin_y)][m3hash] = obj_name;    //tile_origin indexing of the map should help avoiding the race condition
```

Comment:
Either it completely avoids the race condition, or it doesn't: _helping_ avoiding the race condition is not enough :)

In any case, it is completely unsafe to modify a `std::map<>` concurrently so it would first need to be populated with enough entries such that the map itself is never modified.

That said, it seems you don't even need a map at all, you just need an array indexed by tile index (i.e. tile_y * tile_count_y + tile_x).

Reply:
Great tip, thanks. I don't know why I used map of maps here.